File: ciede-2000.lua
-- This function written in Lua is not affiliated with the CIE (International Commission on Illumination),
-- and is released into the public domain. It is provided "as is" without any warranty, express or implied.

-- Classic color accuracy can be ensured with this ΔE, ΔE2000 (ΔE00) implementation.
function ciede_2000(l_1, a_1, b_1, l_2, a_2, b_2)
	-- Working with Lua/LuaJIT and the CIEDE2000 color-difference formula.
	-- k_l, k_c, k_h are parametric factors to be adjusted according to
	-- different viewing parameters such as textures, backgrounds...
	local k_l, k_c, k_h = 1.0, 1.0, 1.0;
	local n = (math.sqrt(a_1 * a_1 + b_1 * b_1) + math.sqrt(a_2 * a_2 + b_2 * b_2)) * 0.5;
	n = n * n * n * n * n * n * n;
	-- A factor involving chroma raised to the power of 7 designed to make
	-- the influence of chroma on the total color difference more accurate.
	n = 1.0 + 0.5 * (1.0 - math.sqrt(n / (n + 6103515625.0)));
	-- hypot from Lua 5.4, rather than sqrt used here can calculate
	-- Euclidean distance while avoiding overflow/underflow.
	local c_1 = math.sqrt(a_1 * a_1 * n * n + b_1 * b_1);
	local c_2 = math.sqrt(a_2 * a_2 * n * n + b_2 * b_2);
	-- atan2 is preferred over atan because it accurately computes the angle of
	-- a point (x, y) in all quadrants, handling the signs of both coordinates.
	local h_1 = math.atan2(b_1, a_1 * n);
	local h_2 = math.atan2(b_2, a_2 * n);
	if h_1 < 0.0 then h_1 = h_1 + 2.0 * math.pi end;
	if h_2 < 0.0 then h_2 = h_2 + 2.0 * math.pi end;
	n = math.abs(h_2 - h_1);
	-- Cross-implementation consistent rounding.
	if math.pi - 1E-14 < n and n < math.pi + 1E-14 then n = math.pi end;
	-- When the hue angles lie in different quadrants, the straightforward
	-- average can produce a mean that incorrectly suggests a hue angle in
	-- the wrong quadrant, the next lines handle this issue.
	local h_m = 0.5 * h_1 + 0.5 * h_2;
	local h_d = (h_2 - h_1) * 0.5;
	if math.pi < n then
		if 0.0 < h_d then
			h_d = h_d - math.pi;
		else
			h_d = h_d + math.pi;
		end
		h_m = h_m + math.pi;
	end
	local p = (36.0 * h_m - 55.0 * math.pi);
	n = (c_1 + c_2) * 0.5;
	n = n * n * n * n * n * n * n;
	-- The hue rotation correction term is designed to account for the
	-- non-linear behavior of hue differences in the blue region.
	local r_t = -2.0 * math.sqrt(n / (n + 6103515625.0)) *
			math.sin(math.pi / 3.0 * math.exp(p * p / (-25.0 * math.pi * math.pi)));
	n = (l_1 + l_2) * 0.5;
	n = (n - 50.0) * (n - 50.0);
	-- Lightness.
	local l = (l_2 - l_1) / (k_l * (1.0 + 0.015 * n / math.sqrt(20.0 + n)));
	-- These coefficients adjust the impact of different harmonic
	-- components on the hue difference calculation.
	local t = 1.0 + 0.24 * math.sin(2.0 * h_m + math.pi / 2.0)
				+ 0.32 * math.sin(3.0 * h_m + 8.0 * math.pi / 15.0)
				- 0.17 * math.sin(h_m + math.pi / 3.0)
				- 0.20 * math.sin(4.0 * h_m + 3.0 * math.pi / 20.0);
	n = c_1 + c_2;
	-- Hue.
	local h = 2.0 * math.sqrt(c_1 * c_2) * math.sin(h_d) / (k_h * (1.0 + 0.0075 * n * t));
	-- Chroma.
	local c = (c_2 - c_1) / (k_c * (1.0 + 0.0225 * n));
	-- Returning the square root ensures that the result represents
	-- the "true" geometric distance in the color space.
	return math.sqrt(l * l + h * h + c * c + c * h * r_t);
end

-- More samples for the CIEDE2000 color difference formula implementation at https://bit.ly/ciede2000-samples
--
-- Testing lab(54.0, -49.0, -78.14) vs lab(54.0, -49.0, -78.14) expects ΔE2000 = ΔE00 = 0.0
-- Testing lab(72.054, 84.93, -37.0) vs lab(72.054, 84.93, -37.077) expects ΔE2000 = ΔE00 = 0.02488428546
-- Testing lab(18.0, -10.4, -93.94) vs lab(18.0, -10.4, -93.754) expects ΔE2000 = ΔE00 = 0.02915457489
-- Testing lab(41.3, -125.6349, 109.49) vs lab(41.3, -126.0, 109.49) expects ΔE2000 = ΔE00 = 0.06307308274
-- Testing lab(94.3, -56.76, -82.37) vs lab(94.3, -56.68, -81.907) expects ΔE2000 = ΔE00 = 0.09335769142
-- Testing lab(12.0, -11.69, 108.46) vs lab(12.0, -12.2, 108.46) expects ΔE2000 = ΔE00 = 0.24122526559
-- Testing lab(63.4718, 91.03, 87.763) vs lab(63.4718, 91.64, 89.7109) expects ΔE2000 = ΔE00 = 0.49909289627
-- Testing lab(67.8081, 66.1529, -90.85) vs lab(67.8081, 66.1529, -88.605) expects ΔE2000 = ΔE00 = 0.78888355437
-- Testing lab(32.6987, 63.4, -38.1) vs lab(32.6987, 66.992, -38.1) expects ΔE2000 = ΔE00 = 1.02227811001
-- Testing lab(88.16, -87.909, 41.0) vs lab(88.16, -87.909, 44.7) expects ΔE2000 = ΔE00 = 1.16160131067
-- Testing lab(22.0, 47.1676, 101.27) vs lab(22.0, 47.1676, 107.0) expects ΔE2000 = ΔE00 = 1.44437521438
-- Testing lab(49.3, -62.72, -63.004) vs lab(49.3, -68.884, -63.004) expects ΔE2000 = ΔE00 = 1.66706610944
-- Testing lab(77.1, 107.7637, 41.0) vs lab(79.729, 107.7637, 41.0) expects ΔE2000 = ΔE00 = 1.850060513
-- Testing lab(103.742, -85.9004, -123.1694) vs lab(107.617, -85.9004, -123.1694) expects ΔE2000 = ΔE00 = 2.11458430274
-- Testing lab(22.445, -53.0, -78.95) vs lab(24.6534, -53.0, -87.6437) expects ΔE2000 = ΔE00 = 2.42886879908
-- Testing lab(6.8724, -126.5158, 33.17) vs lab(11.401, -126.5158, 33.17) expects ΔE2000 = ΔE00 = 2.81399862966
-- Testing lab(125.0, -82.341, 64.121) vs lab(131.2, -82.341, 64.121) expects ΔE2000 = ΔE00 = 2.85769056494
-- Testing lab(91.0, -88.8, 8.6) vs lab(96.41, -88.8, 8.0) expects ΔE2000 = ΔE00 = 3.28401928373
-- Testing lab(8.814, 63.0, -9.885) vs lab(8.814, 63.0, -1.59) expects ΔE2000 = ΔE00 = 3.56140036184
-- Testing lab(93.3712, -117.8581, -88.37) vs lab(99.7, -117.8581, -92.0) expects ΔE2000 = ΔE00 = 3.8118887137
-- Testing lab(40.0, -10.0, 82.087) vs lab(40.0, -16.4182, 77.0) expects ΔE2000 = ΔE00 = 3.93978468698
-- Testing lab(119.2, 103.9, -119.4315) vs lab(127.7, 111.0, -126.1529) expects ΔE2000 = ΔE00 = 4.21411051981
-- Testing lab(122.84, 11.826, 125.0) vs lab(122.84, 2.323, 125.0) expects ΔE2000 = ΔE00 = 4.4637160566
-- Testing lab(93.9, -57.71, 37.0) vs lab(101.9, -57.71, 37.0) expects ΔE2000 = ΔE00 = 4.6636657501
-- Testing lab(81.0, -37.8, 101.1613) vs lab(77.9, -26.42, 99.48) expects ΔE2000 = ΔE00 = 4.99912028709
-- Testing lab(68.0933, -74.5041, -121.5212) vs lab(74.02, -87.0, -119.0) expects ΔE2000 = ΔE00 = 5.3261404711
-- Testing lab(104.62, 51.0, -72.092) vs lab(93.616, 45.0, -72.564) expects ΔE2000 = ΔE00 = 7.02554696063
-- Testing lab(74.104, 60.0852, -82.1848) vs lab(81.24, 82.0, -112.0631) expects ΔE2000 = ΔE00 = 7.67632279584
-- Testing lab(70.0, -82.0, -107.9) vs lab(75.7021, -64.0, -68.6) expects ΔE2000 = ΔE00 = 8.48037477544
-- Testing lab(78.5, -98.7, 28.31) vs lab(88.0, -121.23, 15.894) expects ΔE2000 = ΔE00 = 9.14085757984
-- Testing lab(93.698, -61.708, -127.59) vs lab(82.404, -36.104, -89.1) expects ΔE2000 = ΔE00 = 10.78243115416
-- Testing lab(77.0, 122.7, -103.8) vs lab(83.0421, 83.087, -51.0) expects ΔE2000 = ΔE00 = 11.91398175452
-- Testing lab(23.24, 58.7834, -85.0) vs lab(32.1828, 114.405, -125.818) expects ΔE2000 = ΔE00 = 12.83839415797
-- Testing lab(59.0, -80.88, -82.969) vs lab(47.0, -113.1741, -113.663) expects ΔE2000 = ΔE00 = 13.24144109478
-- Testing lab(89.5233, -34.5, -47.0) vs lab(108.45, -67.9, -80.4) expects ΔE2000 = ΔE00 = 15.00971854267
-- Testing lab(74.665, -58.107, -94.437) vs lab(91.1352, -120.7954, -117.8497) expects ΔE2000 = ΔE00 = 15.9077190011
-- Testing lab(102.4, 41.1, 40.078) vs lab(79.8227, 28.533, 43.9) expects ΔE2000 = ΔE00 = 16.14014008308
-- Testing lab(17.5143, -35.37, 97.82) vs lab(16.7, -91.46, 89.418) expects ΔE2000 = ΔE00 = 17.63516749774
-- Testing lab(113.0, -102.0, -103.72) vs lab(111.1, -25.4, -112.4) expects ΔE2000 = ΔE00 = 19.04293047094
-- Testing lab(96.0, 49.004, -2.224) vs lab(73.6555, 68.33, 25.383) expects ΔE2000 = ΔE00 = 19.76021311016
-- Testing lab(56.0, 115.45, -68.6715) vs lab(35.34, 91.93, -53.831) expects ΔE2000 = ΔE00 = 20.2377752662
-- Testing lab(84.253, 59.96, 9.618) vs lab(93.4426, 43.4, 42.11) expects ΔE2000 = ΔE00 = 21.35888249767
-- Testing lab(36.8, 89.418, 40.98) vs lab(13.2, 122.5, 12.16) expects ΔE2000 = ΔE00 = 22.13950443621
-- Testing lab(73.357, -56.5, -29.8959) vs lab(49.98, -81.883, -79.3) expects ΔE2000 = ΔE00 = 24.03085127618
-- Testing lab(75.856, -114.6, -40.19) vs lab(84.0832, -19.5, -19.68) expects ΔE2000 = ΔE00 = 24.756434996
-- Testing lab(70.0, 70.3, -102.06) vs lab(59.3452, 122.0, 21.5177) expects ΔE2000 = ΔE00 = 38.32041948473
-- Testing lab(31.959, -23.3596, -15.0) vs lab(61.5497, 119.7, 5.0) expects ΔE2000 = ΔE00 = 71.31300475445
-- Testing lab(126.9, -127.9, -92.0) vs lab(20.0, -113.38, -18.5) expects ΔE2000 = ΔE00 = 81.98665602348
-- Testing lab(4.4, 81.603, 66.08) vs lab(76.09, -93.4583, -8.3) expects ΔE2000 = ΔE00 = 101.76517816976
-- Testing lab(118.72, -95.0, -77.74) vs lab(117.81, 90.533, -4.0) expects ΔE2000 = ΔE00 = 128.36286377501

File: ciede-2000.rs
// This function written in Rust is not affiliated with the CIE (International Commission on Illumination),
// and is released into the public domain. It is provided "as is" without any warranty, express or implied.

use std::f64::consts::PI;

// Classic color accuracy can be ensured with this ΔE, ΔE2000 (ΔE00) implementation.
fn ciede_2000(l_1: f64, a_1: f64, b_1: f64, l_2: f64, a_2: f64, b_2: f64) -> f64 {
	// Working with the CIEDE2000 color-difference formula.
	// K_L, K_C, K_H are parametric factors to be adjusted according to
	// different viewing parameters such as textures, backgrounds...
	const K_L: f64 = 1.0;
	const K_C: f64 = 1.0;
	const K_H: f64 = 1.0;
	let mut n = (a_1.hypot(b_1) + a_2.hypot(b_2)) * 0.5;
	n = n * n * n * n * n * n * n;
	// A factor involving chroma raised to the power of 7 designed to make
	// the influence of chroma on the total color difference more accurate.
	n = 1.0 + 0.5 * (1.0 - (n / (n + 6103515625.0)).sqrt());
	// hypot calculates the Euclidean distance while avoiding overflow/underflow.
	let c_1: f64 = (a_1 * n).hypot(b_1);
	let c_2: f64 = (a_2 * n).hypot(b_2);
	// atan2 is preferred over atan because it accurately computes the angle of
	// a point (x, y) in all quadrants, handling the signs of both coordinates.
	let mut h_1 = b_1.atan2(a_1 * n);
	let mut h_2 = b_2.atan2(a_2 * n);
	if h_1 < 0.0 { h_1 += 2.0 * PI; }
	if h_2 < 0.0 { h_2 += 2.0 * PI; }
	n = (h_2 - h_1).abs();
	// Cross-implementation consistent rounding.
	 if (PI - 1e-14..=PI + 1e-14).contains(&n) { n = PI; }
	// When the hue angles lie in different quadrants, the straightforward
	// average can produce a mean that incorrectly suggests a hue angle in
	// the wrong quadrant, the next lines handle this issue.
	let mut h_m = 0.5 * h_1 + 0.5 * h_2;
	let mut h_d = (h_2 - h_1) * 0.5;
	if PI < n {
		if 0.0 < h_d { h_d -= PI; }
		else { h_d += PI; }
		h_m += PI;
	}
	let p = 36.0 * h_m - 55.0 * PI;
	n = (c_1 + c_2) * 0.5;
	n = n * n * n * n * n * n * n;
	// The hue rotation correction term is designed to account for the
	// non-linear behavior of hue differences in the blue region.
	let r_t = -2.0 * (n / (n + 6103515625.0)).sqrt() * (PI / 3.0 * (p * p / (-25.0 * PI * PI)).exp()).sin();
	n = (l_1 + l_2) * 0.5;
	n = (n - 50.0) * (n - 50.0);
	// Lightness.
	let l = (l_2 - l_1) / (K_L * (1.0 + 0.015 * n / (20.0 + n).sqrt()));
	// These coefficients adjust the impact of different harmonic
	// components on the hue difference calculation.
	let t = 1.0 	+ 0.24 * (2.0 * h_m + PI * 0.5).sin()
			+ 0.32 * (3.0 * h_m + 8.0 * PI / 15.0).sin()
			- 0.17 * (h_m + PI / 3.0).sin()
			- 0.20 * (4.0 * h_m + 3.0 * PI / 20.0).sin();
	n = c_1 + c_2;
	// Hue.
	let h = 2.0 * (c_1 * c_2).sqrt() * (h_d).sin() / (K_H * (1.0 + 0.0075 * n * t));
	// Chroma.
	let c = (c_2 - c_1) / (K_C * (1.0 + 0.0225 * n));
	// Returning the square root ensures that the result represents
	// the "true" geometric distance in the color space.
	(l * l + h * h + c * c + c * h * r_t).sqrt()
}

//
// More samples for the CIEDE2000 color difference formula implementation at https://bit.ly/ciede2000-samples
//
// Testing lab(128.0, -65.44, -8.3) vs lab(128.0, -65.44, -8.3) expects ΔE2000 = ΔE00 = 0.0
// Testing lab(8.8936, -102.8, 86.8) vs lab(8.8936, -102.8, 86.849) expects ΔE2000 = ΔE00 = 0.01088010727
// Testing lab(42.9, 53.0, -50.9) vs lab(42.9, 53.0, -51.0) expects ΔE2000 = ΔE00 = 0.0409705579
// Testing lab(14.9798, -70.632, 82.2) vs lab(14.9798, -70.632, 82.4934) expects ΔE2000 = ΔE00 = 0.07260054571
// Testing lab(54.0, -100.45, 23.903) vs lab(54.0, -100.9037, 23.903) expects ΔE2000 = ΔE00 = 0.08653303132
// Testing lab(40.0, 80.647, -4.0) vs lab(40.0, 79.4, -4.0) expects ΔE2000 = ΔE00 = 0.27145472641
// Testing lab(36.2134, 108.0, 43.8) vs lab(36.2134, 108.0, 42.586) expects ΔE2000 = ΔE00 = 0.437185933
// Testing lab(46.281, 9.9409, 48.1) vs lab(46.281, 9.0, 48.1) expects ΔE2000 = ΔE00 = 0.65133785804
// Testing lab(40.484, 126.842, 104.339) vs lab(40.484, 122.74, 104.339) expects ΔE2000 = ΔE00 = 1.03360227874
// Testing lab(44.0, -123.9805, 114.45) vs lab(44.0, -130.94, 114.0) expects ΔE2000 = ΔE00 = 1.24008633516
// Testing lab(105.0, 92.0, 92.142) vs lab(105.0, 87.59, 92.142) expects ΔE2000 = ΔE00 = 1.44597440672
// Testing lab(17.8958, -60.74, -9.44) vs lab(17.8958, -60.74, -6.119) expects ΔE2000 = ΔE00 = 1.77058503931
// Testing lab(126.0, -54.91, -2.354) vs lab(126.0, -56.7, -5.9) expects ΔE2000 = ΔE00 = 1.98149019821
// Testing lab(80.99, -2.2, 78.0) vs lab(80.99, -2.2, 68.455) expects ΔE2000 = ΔE00 = 2.22692613885
// Testing lab(105.9, 17.534, -93.3) vs lab(108.86, 21.0, -94.0) expects ΔE2000 = ΔE00 = 2.40280987434
// Testing lab(84.3, -84.61, 69.1) vs lab(88.083, -84.61, 65.4484) expects ΔE2000 = ΔE00 = 2.64030590341
// Testing lab(93.8774, 26.93, 36.21) vs lab(93.8774, 22.0, 36.21) expects ΔE2000 = ΔE00 = 3.04909996041
// Testing lab(108.51, 123.37, 80.5698) vs lab(108.51, 123.37, 70.6) expects ΔE2000 = ΔE00 = 3.20639516185
// Testing lab(104.649, 9.759, -33.0) vs lab(111.192, 9.759, -33.0) expects ΔE2000 = ΔE00 = 3.50600010455
// Testing lab(111.46, 49.901, 38.0) vs lab(111.46, 55.1911, 48.0) expects ΔE2000 = ΔE00 = 3.61485062757
// Testing lab(70.414, -83.48, 4.1603) vs lab(70.414, -83.48, 13.29) expects ΔE2000 = ΔE00 = 3.91631093378
// Testing lab(120.8, -57.333, -88.89) vs lab(129.32, -51.991, -88.89) expects ΔE2000 = ΔE00 = 4.27118639761
// Testing lab(123.31, 124.8963, 66.396) vs lab(132.76, 124.8963, 66.396) expects ΔE2000 = ΔE00 = 4.35763501084
// Testing lab(121.8, -7.91, 44.76) vs lab(123.0, -15.9, 46.9) expects ΔE2000 = ΔE00 = 4.83954494008
// Testing lab(120.8, -84.1097, -98.602) vs lab(114.2217, -76.0, -113.6) expects ΔE2000 = ΔE00 = 5.03282590435
// Testing lab(4.9657, 95.0, -84.772) vs lab(11.0, 120.3723, -98.15) expects ΔE2000 = ΔE00 = 5.62650839731
// Testing lab(35.5064, 99.0, 63.0) vs lab(36.049, 110.0, 86.8606) expects ΔE2000 = ΔE00 = 6.31169733828
// Testing lab(53.08, -79.0, -126.4543) vs lab(58.3689, -82.06, -96.0) expects ΔE2000 = ΔE00 = 7.45551466404
// Testing lab(124.91, -74.49, -99.5) vs lab(114.041, -97.816, -86.839) expects ΔE2000 = ΔE00 = 8.51254802065
// Testing lab(78.2445, 51.773, 73.7) vs lab(86.5, 82.0, 101.3076) expects ΔE2000 = ΔE00 = 9.46437310384
// Testing lab(39.0, -55.043, -81.443) vs lab(44.9647, -35.0, -109.73) expects ΔE2000 = ΔE00 = 10.28718759581
// Testing lab(14.8, 54.73, -112.37) vs lab(25.67, 18.075, -63.8992) expects ΔE2000 = ΔE00 = 11.85797599414
// Testing lab(114.726, -121.77, 85.5) vs lab(114.0, -73.856, 102.162) expects ΔE2000 = ΔE00 = 12.46634471848
// Testing lab(95.9508, 81.5, 96.0) vs lab(94.861, 45.0969, 89.163) expects ΔE2000 = ΔE00 = 13.50389100707
// Testing lab(53.9, -5.9, 73.0) vs lab(42.904, 8.8451, 88.51) expects ΔE2000 = ΔE00 = 14.1273649909
// Testing lab(37.4, -44.732, 85.444) vs lab(52.8, -48.0, 66.6) expects ΔE2000 = ΔE00 = 15.71728968186
// Testing lab(40.9, -27.743, -122.0544) vs lab(27.4, -7.035, -83.0) expects ΔE2000 = ΔE00 = 16.14982484415
// Testing lab(79.28, 88.0, -72.521) vs lab(104.0, 99.8, -50.52) expects ΔE2000 = ΔE00 = 17.33682468253
// Testing lab(46.0, -122.0, -115.9) vs lab(32.48, -93.0, -44.5) expects ΔE2000 = ΔE00 = 18.64194324706
// Testing lab(89.3484, 113.917, 21.77) vs lab(84.78, 44.7578, -11.7881) expects ΔE2000 = ΔE00 = 19.59176199882
// Testing lab(77.5, -108.58, 126.0) vs lab(105.05, -72.175, 60.436) expects ΔE2000 = ΔE00 = 20.80923048858
// Testing lab(70.9, -116.0, -72.253) vs lab(54.6383, -51.879, -89.0) expects ΔE2000 = ΔE00 = 21.35609387699
// Testing lab(10.33, -9.009, -93.0) vs lab(4.67, 32.7, -93.5) expects ΔE2000 = ΔE00 = 22.82541279255
// Testing lab(99.715, -98.892, 123.37) vs lab(88.05, -120.184, 31.07) expects ΔE2000 = ΔE00 = 24.02279121533
// Testing lab(95.2, 49.8454, 76.5) vs lab(84.9099, 12.074, 100.702) expects ΔE2000 = ΔE00 = 24.58449512986
// Testing lab(35.0346, -80.438, 8.255) vs lab(61.6, -20.0, -86.15) expects ΔE2000 = ΔE00 = 49.20890735575
// Testing lab(80.0, -88.78, -77.5569) vs lab(79.479, 66.3, -104.19) expects ΔE2000 = ΔE00 = 55.4456271891
// Testing lab(59.0, 93.545, -8.0) vs lab(114.863, -26.1738, -43.0) expects ΔE2000 = ΔE00 = 85.01403243966
// Testing lab(45.8, -51.79, -117.35) vs lab(115.842, 111.1864, -36.4) expects ΔE2000 = ΔE00 = 118.44269488166
// Testing lab(101.4, 119.5, 13.24) vs lab(5.76, -37.965, -110.1606) expects ΔE2000 = ΔE00 = 130.70704766893
